pkg/crypto/certloader/cell.go

```diff
@@ -39,37 +38,95 @@ type Config struct {
 // and the promise provided wrapped in another type to avoid possible conflicts/replacements
 // when used multiple times in the same hive.
 func NewWatchedServerConfigPromise(lc cell.Lifecycle, jobGroup job.Group, log *slog.Logger, cfg Config) (promise.Promise[*WatchedServerConfig], error) {
-	log = log.With(logfields.Config, "certloader-tls")
+	log = log.With(logfields.Config, "certloader-server-tls")
 	if !cfg.TLS {
 		log.Info("Certloader TLS watcher disabled")
 		return nil, nil
 	}
 
 	resolver, promise := promise.New[*WatchedServerConfig]()
 
-	jobGroup.Add(job.OneShot("certloader-tls", func(ctx context.Context, _ cell.Health) error {
-		metricsTLSConfigChan, err := FutureWatchedServerConfig(
+	jobGroup.Add(job.OneShot("certloader-server-tls", func(ctx context.Context, _ cell.Health) error {
+		watchedConfigChan, err := FutureWatchedServerConfig(
 			ctx, log,
 			cfg.TLSClientCAFiles, cfg.TLSCertFile, cfg.TLSKeyFile,
 		)
 		if err != nil {
-			err := fmt.Errorf("failed to initialize certloader TLS configuration: %w", err)
+			err := fmt.Errorf("failed to initialize certloader TLS watched server configuration: %w", err)
 			resolver.Reject(err)
 			return err
 		}
 
 		waitingMsgTimeout := time.After(30 * time.Second)
-		var metricsTLSConfig *WatchedServerConfig
-		for metricsTLSConfig == nil {
+		var watchedConfig *WatchedServerConfig
+		for watchedConfig == nil {
 			select {
-			case metricsTLSConfig = <-metricsTLSConfigChan:
+			case watchedConfig = <-watchedConfigChan:
 			case <-waitingMsgTimeout:
 				log.Info("Waiting for certloader TLS certificate and key files to be created")
 			case <-ctx.Done():
 				return nil
 			}
 		}
-		resolver.Resolve(metricsTLSConfig)
+		resolver.Resolve(watchedConfig)
+		return nil
+	}))
+
+	lc.Append(cell.Hook{
+		OnStop: func(ctx cell.HookContext) error {
+			// stop the resolved config watcher (best effort)
+			ctx, cancel := context.WithTimeout(ctx, time.Second)
+			defer cancel()
+			cfg, _ := promise.Await(ctx)
+			if cfg != nil {
+				cfg.Stop()
+			}
+			return nil
+		},
+	})
+
+	return promise, nil
+}
+
+// NewWatchedClientConfigPromise provides a promise that resolves to a WatchedClientConfig.
+// The resolved config can be used to obtain a tls.Config. The promise resolves when the
+// certificates become ready and have been loaded.
+//
+// This is meant to be used as a Hive constructor and is recommended to be placed in a module
+// and the promise provided wrapped in another type to avoid possible conflicts/replacements
+// when used multiple times in the same hive.
+func NewWatchedClientConfigPromise(lc cell.Lifecycle, jobGroup job.Group, log *slog.Logger, cfg Config) (promise.Promise[*WatchedClientConfig], error) {
```

Comment:
This is a copy-pasta of `NewWatchedServerConfigPromise` as I couldn't figure out a way to merge both into an abstraction without also merging the `ClientConnBuilder` and `ServerConnBuilder` interfaces. I think this could be done in a straightforward way but didn't want to attempt it all in the same PR.

Reply:
My initial reaction was that this should be easy to de-duplicate, but you're right, it's not quite as straight forward. This is fine,  let's keep it WET.